jib-core/src/test/java/com/google/cloud/tools/jib/image/LayerBuilderTest.java

```diff
@@ -96,19 +119,54 @@ public void testBuild() throws URISyntaxException, IOException {
                   throw new RuntimeException(ex);
                 }
               });
+    }
+  }
 
-      // Verifies that blobA was added.
-      TarArchiveEntry header = tarArchiveInputStream.getNextTarEntry();
-      String expectedExtractionPath = "extract/here/blobA";
-      Assert.assertEquals(expectedExtractionPath, header.getName());
-
-      String expectedFileString = new String(Files.readAllBytes(blobA), StandardCharsets.UTF_8);
-
-      String extractedFileString =
-          CharStreams.toString(
-              new InputStreamReader(tarArchiveInputStream, StandardCharsets.UTF_8));
+  @Test
+  public void testToBlob_reproducibility()
+      throws IOException, NoSuchAlgorithmException, InterruptedException {
+    Path testRoot = temporaryFolder.getRoot().toPath();
+    Path root1 = Files.createDirectories(testRoot.resolve("files1"));
+    Path root2 = Files.createDirectories(testRoot.resolve("files2"));
+
+    String extractionPath = "/somewhere";
+
+    String contentA = "abcabc";
+    Path fileA1 = createFile(root1, "fileA", contentA, 10000);
+    Path fileA2 = createFile(root2, "fileA", contentA, 20000);
+    String contentB = "yumyum";
+    Path fileB1 = createFile(root1, "fileB", contentB, 10000);
+    Path fileB2 = createFile(root2, "fileB", contentB, 20000);
+
+    // check if modified times are off
+    Assert.assertNotEquals(Files.getLastModifiedTime(fileA1), Files.getLastModifiedTime(fileA2));
+    Assert.assertNotEquals(Files.getLastModifiedTime(fileB1), Files.getLastModifiedTime(fileB2));
+
+    // create layers of exact same content but ordered differently and with different timestamps
+    Blob layer =
+        new LayerBuilder(Arrays.asList(fileA1, fileB1), extractionPath, true).build().getBlob();
+    Blob reproduced =
+        new LayerBuilder(Arrays.asList(fileB2, fileA2), extractionPath, true).build().getBlob();
+
+    // this is a control layer that is the same as 'layerReproduced' without reproducibility on
+    Blob notReproduced =
+        new LayerBuilder(Arrays.asList(fileB2, fileA2), extractionPath, false).build().getBlob();
+
+    MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
```

Comment:
There's the CountingDigestOutputStream that is used in other tests for generating layer hashes.

For example:

```java
CountingDigestOutputStream countingDigestOutputStream = new CountingDigestOutputStream(ByteStreams.nullOutputStream())
layer.writeTo(countingDigestOutputStream);
BlobDescriptor layerBlobDescriptor = countingDigestOutputStream.toBlobDescriptor();
```

Although, here it seems that it might be simpler to just check the equality of `Blobs.writeToString()`?

Reply:
Actually, it might not be completely safe to write the tar gzip BLOB to a string and check that. Perhaps it'd be better to `blob.writeTo` some byte array and then directly compare those?